File: kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/definition/v1_1/ContextEntry.java
/*
 * Copyright 2017 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.kie.workbench.common.dmn.api.definition.v1_1;

import java.util.ArrayList;
import java.util.List;

import org.jboss.errai.common.client.api.annotations.Portable;
import org.kie.workbench.common.dmn.api.definition.HasExpression;
import org.kie.workbench.common.dmn.api.definition.HasTypeRef;
import org.kie.workbench.common.dmn.api.definition.HasTypeRefs;
import org.kie.workbench.common.dmn.api.definition.HasVariable;
import org.kie.workbench.common.stunner.core.util.HashUtil;

import static org.kie.workbench.common.dmn.api.definition.v1_1.common.HasTypeRefHelper.getNotNullHasTypeRefs;

@Portable
public class ContextEntry extends DMNModelInstrumentedBase implements HasExpression,
                                                                      HasTypeRefs,
                                                                      HasVariable<InformationItem> {

    private InformationItem variable;
    private Expression expression;

    @Override
    public InformationItem getVariable() {
        return variable;
    }

    @Override
    public void setVariable(final InformationItem value) {
        this.variable = value;
    }

    @Override
    public Expression getExpression() {
        return expression;
    }

    @Override
    public void setExpression(final Expression value) {
        this.expression = value;
    }

    @Override
    public DMNModelInstrumentedBase asDMNModelInstrumentedBase() {
        return this;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof ContextEntry)) {
            return false;
        }

        final ContextEntry that = (ContextEntry) o;

        if (variable != null ? !variable.equals(that.variable) : that.variable != null) {
            return false;
        }
        return expression != null ? expression.equals(that.expression) : that.expression == null;
    }

    @Override
    public int hashCode() {
        return HashUtil.combineHashCodes(variable != null ? variable.hashCode() : 0,
                                         expression != null ? expression.hashCode() : 0);
    }

    @Override
    public List<HasTypeRef> getHasTypeRefs() {

        final List<HasTypeRef> hasTypeRefs = new ArrayList<>();

        hasTypeRefs.addAll(getNotNullHasTypeRefs(getExpression()));
        hasTypeRefs.addAll(getNotNullHasTypeRefs(getVariable()));

        return hasTypeRefs;
    }
}

File: kogito-editors-js/packages/boxed-expression-component/src/api/ContextEntry.ts
/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DataType } from "./DataType";
import { ExpressionProps } from "./ExpressionProps";
import * as _ from "lodash";
import { DataRecord, Row } from "react-table";
import { TableHandlerConfiguration, TableOperation } from "./Table";
import { BoxedExpressionEditorI18n } from "../i18n";

export interface EntryInfo {
  /** Entry name */
  name: string;
  /** Entry data type */
  dataType: DataType;
}

export interface ContextEntryRecord {
  entryInfo: EntryInfo;
  /** Entry expression */
  entryExpression: ExpressionProps;
  /** Label used for the popover triggered when editing info section */
  editInfoPopoverLabel?: string;
  /** True, for synchronizing name and dataType parameters, between entryInfo and entryExpression */
  nameAndDataTypeSynchronized?: boolean;
  /** Callback to be invoked on expression resetting */
  onExpressionResetting?: () => void;
}

export type ContextEntries = ContextEntryRecord[];

export const DEFAULT_ENTRY_INFO_MIN_WIDTH = 150;
export const DEFAULT_ENTRY_EXPRESSION_MIN_WIDTH = 370;

export const getHandlerConfiguration = (
  i18n: BoxedExpressionEditorI18n,
  groupName: string
): TableHandlerConfiguration => [
  {
    group: groupName,
    items: [
      { name: i18n.rowOperations.insertAbove, type: TableOperation.RowInsertAbove },
      { name: i18n.rowOperations.insertBelow, type: TableOperation.RowInsertBelow },
      { name: i18n.rowOperations.delete, type: TableOperation.RowDelete },
      { name: i18n.rowOperations.clear, type: TableOperation.RowClear },
    ],
  },
];

export const generateNextAvailableEntryName = (
  entryInfos: EntryInfo[],
  namePrefix: string,
  lastIndex: number = entryInfos.length
): string => {
  const candidateName = `${namePrefix}-${lastIndex === 0 ? 1 : lastIndex}`;
  const entryWithCandidateName = _.find(entryInfos, { name: candidateName });
  return entryWithCandidateName ? generateNextAvailableEntryName(entryInfos, namePrefix, lastIndex + 1) : candidateName;
};

export const getEntryKey = (row: Row): string => {
  const entryRecord = row.original as ContextEntryRecord;
  return entryRecord.entryInfo.name + entryRecord.entryInfo.dataType;
};

export const resetEntry = (row: DataRecord): DataRecord => ({
  ...row,
  entryExpression: { uid: (row.entryExpression as ExpressionProps).uid },
});
